### PR TITLE
Add RealmObject.addListener()

### LINKED ISSUE #763: Ability to observe individual objects
I am big fan of both realm and react native and realm has made me love writing apps even more.
If only there was a thing like a listener on individual object, it would have been great. 
Also, should I expect anything like this in near future?
I have seen this discussion on some other thread too, but in a nested object, it doesn't send out notifications on modifying an object of a list child. 
There could both good and bad parts to it. AFAIK, android has it, do we plan to have for JS too?

Thanks for this great library.

### PR DESCRIPTION
**Pull request by kneth**
## What, How & Why?

This closes #763

## ☑️ ToDos

* [x] 📝 Changelog entry
* [x] 📝 `Compatibility` label is updated or copied from previous entry
* [x] 🚦 Tests
* ~[ ] 📝 Public documentation PR created or is not necessary~
* ~[ ] 💥 `Breaking` label has been applied or is not necessary~

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [x] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

**Comment by kneth** [comment]
I am a bit concerned about the Chrome debug API but the rest is fine.

@bmunkholm @cmelchior Please review

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 x.x.x Release notes (yyyy-MM-dd)
 =============================================================
 ### Enhancements
+* Added `RealmObject.addListener()` to set up object-level notifications. ([#763](https://github.com/realm/realm-js/issues/763))
```
Comment (cmelchior) [inline CHANGELOG.md]:
```suggestion
* Added `RealmObject.addListener()` and `RealmObject.removeListener()` to set up and remove object-level notifications. ([#763](https://github.com/realm/realm-js/issues/763))
```

File: docs/object.js [hunk 2]
```diff
@@ -54,4 +54,36 @@ class Object {
...
      */
     linkingObjectsCount() {}
+
+    /**
+     * Add a listener `callback` which will be called when a **live** object instance changes.
+     * @param {function(obj, changes)} callback - A function to be called when changes occur.
+     *   The callback function is called with two arguments:
+     *   - `obj`: the object that changed,
+     *   - `changes`: a dictionary with keys `deleted`, and `changedProperties`. `deleted` is true
+     *       if the object has been deleted. `changesProperties` is an array of property names of changes
```
Comment (cmelchior) [inline docs/object.js]:
```suggestion
     *       if the object has been deleted. `changesProperties` is an array of property that has changed
```

File: docs/object.js [hunk 3]
```diff
@@ -54,4 +54,36 @@ class Object {
...
+     *       properties.
```
Comment (cmelchior) [inline docs/object.js]:
```suggestion
     *       their value. 
```

File: lib/index.d.ts [hunk 4]
```diff
@@ -100,6 +110,8 @@ declare namespace Realm {
         [keys: string]: any;
     }
 
+    type ObjectChangeCallback = (object: Object, changes: any) => void;
```
Comment (cmelchior) [inline lib/index.d.ts]: Shouldn't we create an `ObjectChangeSet` interface here instead of `any` ?

Comment (kneth) [reply lib/index.d.ts]: true

File: docs/object.js [hunk 5]
```diff
@@ -54,4 +54,36 @@ class Object {
...
      */
     linkingObjectsCount() {}
+
+    /**
+     * Add a listener `callback` which will be called when a **live** object instance changes.
+     * @param {function(obj, changes)} callback - A function to be called when changes occur.
+     *   The callback function is called with two arguments:
+     *   - `obj`: the object that changed,
+     *   - `changes`: a dictionary with keys `deleted`, and `changedProperties`. `deleted` is true
+     *       if the object has been deleted. `changesProperties` is an array of property that has changed
```
Comment (bmunkholm) [inline docs/object.js]: property -> properties

File: docs/object.js [hunk 6]
```diff
@@ -54,4 +54,36 @@ class Object {
...
+     *       their value. 
+     * @throws {Error} If `callback` is not a function.
+     * @since 2.23.0
+     * @example
+     * wine.addListener((obj, changes) => {
+     *  // obj === wine
+     *  console.log(`object is deleted: ${changes.deleted}`);
+     *  console.log(`${changes.changedProperties.length} properties have been changed`);
+     * })
```
Comment (bmunkholm) [inline docs/object.js]: perhaps iterate over them and log their names?

File: tests/js/object-tests.js [hunk 7]
```diff
@@ -562,5 +562,46 @@ module.exports = {
             obj._setLink('stringLink', 'c');
             TestCase.assertEqual(obj.stringLink, null);
         });
+    },
+
+    testNotification: function() {
```
Comment (bmunkholm) [inline tests/js/object-tests.js]: missing tests of removeListener

**Comment by bmunkholm** [comment]
I suggest having @tgoyne review this as well

File: docs/object.js [hunk 8]
```diff
@@ -54,4 +54,25 @@ class Object {
      * @since 2.6.0
      */
     linkingObjectsCount() {}
+
+    /**
+     * Add a listener `callback` which will be called when a **live** object instance changes.
+     * @param {function(collection, changes)} callback - A function to be called when changes occur.
```
[the remark this answers is not in the document]
Comment (kneth) [reply docs/object.js]: I have updated it to `obj`

**Pull request merged**